In [2]:
import pandas as pd
from pathlib import Path

DATA_DIR = Path("../../data")  # FH/data relative to FH/AI/notebooks/

files = {
    "sanity": DATA_DIR / "FOCUS_Data_Export_TABLE_NW_SANITY.xlsx",
    "rxlev":  DATA_DIR / "FOCUS_Data_Export_TABLE_NW_RXLev.xlsx",
    "fh_rsl": DATA_DIR / "temp" / "export.xlsx",
    "rtwp":   DATA_DIR / "temp" / "export_optim.xlsx",
}

for name, path in files.items():
    print(name, "->", path.resolve(), "| exists:", path.exists())

sanity -> C:\OrangeCopilot\projet_pfe-main (1)\projet_pfe-main\main\FH\data\FOCUS_Data_Export_TABLE_NW_SANITY.xlsx | exists: True
rxlev -> C:\OrangeCopilot\projet_pfe-main (1)\projet_pfe-main\main\FH\data\FOCUS_Data_Export_TABLE_NW_RXLev.xlsx | exists: True
fh_rsl -> C:\OrangeCopilot\projet_pfe-main (1)\projet_pfe-main\main\FH\data\temp\export.xlsx | exists: True
rtwp -> C:\OrangeCopilot\projet_pfe-main (1)\projet_pfe-main\main\FH\data\temp\export_optim.xlsx | exists: True


Cell 2 — load all four files:

In [3]:
dfs = {name: pd.read_excel(path) for name, path in files.items()}
print("loaded:", list(dfs.keys()))

loaded: ['sanity', 'rxlev', 'fh_rsl', 'rtwp']


Cell 3 — shape and dtypes:

In [4]:
for name, df in dfs.items():
    print(f"--- {name} ---")
    print("shape:", df.shape)
    print(df.dtypes)
    print()

--- sanity ---
shape: (6039, 14)
FOCUS_Data_Export_TABLE_NW_SANITY       str
Unnamed: 1                           object
Unnamed: 2                              str
Unnamed: 3                              str
Unnamed: 4                              str
Unnamed: 5                           object
Unnamed: 6                           object
Unnamed: 7                           object
Unnamed: 8                           object
Unnamed: 9                           object
Unnamed: 10                          object
Unnamed: 11                             str
Unnamed: 12                             str
Unnamed: 13                             str
dtype: object

--- rxlev ---
shape: (6017, 6)
FOCUS_Data_Export_TABLE_NW_RXLev       str
Unnamed: 1                             str
Unnamed: 2                          object
Unnamed: 3                             str
Unnamed: 4                             str
Unnamed: 5                             str
dtype: object

--- fh_rsl ---
shape: (1292, 8)


Cell 4 — head preview:

In [5]:
for name, df in dfs.items():
    print(f"--- {name} ---")
    display(df.head())

--- sanity ---


,FOCUS_Data_Export_TABLE_NW_SANITY,Unnamed: 1,Unnamed: 2,Unnamed: 3,Unnamed: 4,Unnamed: 5,Unnamed: 6,Unnamed: 7,Unnamed: 8,Unnamed: 9,Unnamed: 10,Unnamed: 11,Unnamed: 12,Unnamed: 13
0,IP,Slot,Sanity,Mod (Ref),Mod (Min),UAS,SEP,SES,ES,BBE,OFS,RSL (Min),RSL (Max),RSL (Avg)
1,172.18.172.39,1,CURATIVE,2048 QAM,Unknown,691200,0,0,0,0,95,-93.80 dBm,-88.20 dBm,-92.06 dBm
2,172.19.228.26,2,CURATIVE,256 QAM,Unknown,691200,0,0,0,0,570618,-99.90 dBm,-71.00 dBm,-82.62 dBm
3,172.18.4.56,1,CURATIVE,128 QAM,Unknown,691200,0,0,0,0,7209,-90.90 dBm,-57.80 dBm,-70.39 dBm
4,172.18.10.3,1,CURATIVE,64 QAM,Unknown,691162,0,0,0,0,1,-91.10 dBm,-29.40 dBm,-77.59 dBm


--- rxlev ---


,FOCUS_Data_Export_TABLE_NW_RXLev,Unnamed: 1,Unnamed: 2,Unnamed: 3,Unnamed: 4,Unnamed: 5
0,Name,IP,Slot,Min RSL,Avg RSL,Max RSL
1,SFA_0112_SBO_0015_SFA_0330,172.19.34.36,5,-99.90 dBm,-87.56 dBm,-40.40 dBm
2,ARI_0166_JUMIA_SERV_CTF_Charguia,172.19.132.17,2,-99.90 dBm,-86.59 dBm,-32.60 dBm
3,JEN_0038_JEN_0003,172.19.228.26,2,-99.90 dBm,-82.62 dBm,-71.00 dBm
4,SFA_0104_SFA_105,172.18.46.47,4,-99.90 dBm,-79.75 dBm,-30.70 dBm


--- fh_rsl ---


,Status,IP,Max RSL,RSL REF,RSL DIFF,Name,File,Comment
0,Lien_OK,172.19.218.26,-26.6,-30.88,-4.28,B2B_TTDE_TUN_0058,TTDE_TUN_0058.xlsx,NaN
1,Lien_OK,172.18.128.4,-38.4,-40.20,-1.80,B2B_ATB_KELIBIA_NAB_0061,ATB_KELIBIA_NAB_0061.xlsx,NaN
2,Lien_OK,172.18.22.19,-35.6,-35.97,-0.37,B2B_ATB_CHOTRANA_ARI_0092,ATB_CHOTRANA_ARI_0092.xlsx,NaN
3,Lien_OK,172.18.46.130,-35.3,-37.41,-2.11,B2B_ISB_SFA_0001,B2B_ISB_SFA_0001.xlsx,NaN
4,Lien_OK,172.18.70.12,-30.5,-32.20,-1.70,B2B_BT_MORNEG_BAR_0046,BT_MORNEG_BAR_0046.xlsx,NaN


--- rtwp ---


,Cell_Name,mean_rtwp,max_rtwp,min_rtwp,occ_rtwp_gt_98,vs_mean_rtwp_ant0,vs_mean_rtwp_ant1,occ_diff_sect_gt_5
0,ARI_0017_C04_309_w3,-100.7470,-90.1283,-104.6680,14,-100.7670,-99.1217,0
1,ARI_0025_C03_309_w4,-97.7750,-93.3045,-99.4646,34,-99.9746,-96.3378,9
2,ARI_0029_C01_309_w3,-99.6981,-92.3193,-102.5130,12,-98.8377,-98.7916,0
3,ARI_0033_C09_309_w3,-99.5026,-82.7024,-102.9390,13,-98.6392,-98.6568,0
4,ARI_0034_C02_309_v8,-101.2600,-84.9180,-105.4310,15,-100.3930,-100.6230,0


Cell 5 — missing values and duplicates:

In [6]:
for name, df in dfs.items():
    print(f"--- {name} ---")
    print("missing values:\n", df.isnull().sum())
    print("duplicate rows:", df.duplicated().sum())
    print()

--- sanity ---
missing values:
 FOCUS_Data_Export_TABLE_NW_SANITY    0
Unnamed: 1                           0
Unnamed: 2                           0
Unnamed: 3                           0
Unnamed: 4                           0
Unnamed: 5                           0
Unnamed: 6                           0
Unnamed: 7                           0
Unnamed: 8                           0
Unnamed: 9                           0
Unnamed: 10                          0
Unnamed: 11                          3
Unnamed: 12                          3
Unnamed: 13                          3
dtype: int64
duplicate rows: 0

--- rxlev ---
missing values:
 FOCUS_Data_Export_TABLE_NW_RXLev    0
Unnamed: 1                          0
Unnamed: 2                          0
Unnamed: 3                          0
Unnamed: 4                          0
Unnamed: 5                          0
dtype: int64
duplicate rows: 0

--- fh_rsl ---
missing values:
 Status         0
IP             0
Max RSL        0
RSL REF        0

cell 6 : 

In [7]:
# reload sanity and rxlev with the correct header row
dfs["sanity"] = pd.read_excel(files["sanity"], header=1)
dfs["rxlev"] = pd.read_excel(files["rxlev"], header=1)

for name in ["sanity", "rxlev"]:
    print(f"--- {name} ---")
    print("shape:", dfs[name].shape)
    print(dfs[name].columns.tolist())
    display(dfs[name].head())
    print()

--- sanity ---
shape: (6038, 14)
['IP', 'Slot', 'Sanity', 'Mod (Ref)', 'Mod (Min)', 'UAS', 'SEP', 'SES', 'ES', 'BBE', 'OFS', 'RSL (Min)', 'RSL (Max)', 'RSL (Avg)']


,IP,Slot,Sanity,Mod (Ref),Mod (Min),UAS,SEP,SES,ES,BBE,OFS,RSL (Min),RSL (Max),RSL (Avg)
0,172.18.172.39,1,CURATIVE,2048 QAM,Unknown,691200,0,0,0,0,95,-93.80 dBm,-88.20 dBm,-92.06 dBm
1,172.19.228.26,2,CURATIVE,256 QAM,Unknown,691200,0,0,0,0,570618,-99.90 dBm,-71.00 dBm,-82.62 dBm
2,172.18.4.56,1,CURATIVE,128 QAM,Unknown,691200,0,0,0,0,7209,-90.90 dBm,-57.80 dBm,-70.39 dBm
3,172.18.10.3,1,CURATIVE,64 QAM,Unknown,691162,0,0,0,0,1,-91.10 dBm,-29.40 dBm,-77.59 dBm
4,172.18.96.16,2,CURATIVE,512 QAM,Unknown,604800,0,0,0,0,352600,-92.80 dBm,-23.80 dBm,-67.34 dBm



--- rxlev ---
shape: (6016, 6)
['Name', 'IP', 'Slot', 'Min RSL', 'Avg RSL', 'Max RSL']


,Name,IP,Slot,Min RSL,Avg RSL,Max RSL
0,SFA_0112_SBO_0015_SFA_0330,172.19.34.36,5,-99.90 dBm,-87.56 dBm,-40.40 dBm
1,ARI_0166_JUMIA_SERV_CTF_Charguia,172.19.132.17,2,-99.90 dBm,-86.59 dBm,-32.60 dBm
2,JEN_0038_JEN_0003,172.19.228.26,2,-99.90 dBm,-82.62 dBm,-71.00 dBm
3,SFA_0104_SFA_105,172.18.46.47,4,-99.90 dBm,-79.75 dBm,-30.70 dBm
4,TUN_0092_FRANCHISE_MOUROUJ1,172.18.8.1,1,-99.90 dBm,-77.67 dBm,-53.90 dBm


In [8]:
for name in ["sanity", "rxlev", "fh_rsl", "rtwp"]:
    print(f"--- {name} ---")
    print(dfs[name].dtypes)
    print("missing values:\n", dfs[name].isnull().sum())
    print("duplicate rows:", dfs[name].duplicated().sum())
    print()

--- sanity ---
IP             str
Slot         int64
Sanity         str
Mod (Ref)      str
Mod (Min)      str
UAS          int64
SEP          int64
SES          int64
ES           int64
BBE          int64
OFS          int64
RSL (Min)      str
RSL (Max)      str
RSL (Avg)      str
dtype: object
missing values:
 IP           0
Slot         0
Sanity       0
Mod (Ref)    0
Mod (Min)    0
UAS          0
SEP          0
SES          0
ES           0
BBE          0
OFS          0
RSL (Min)    3
RSL (Max)    3
RSL (Avg)    3
dtype: int64
duplicate rows: 0

--- rxlev ---
Name         str
IP           str
Slot       int64
Min RSL      str
Avg RSL      str
Max RSL      str
dtype: object
missing values:
 Name       0
IP         0
Slot       0
Min RSL    0
Avg RSL    0
Max RSL    0
dtype: int64
duplicate rows: 0

--- fh_rsl ---
Status          str
IP              str
Max RSL     float64
RSL REF     float64
RSL DIFF    float64
Name            str
File            str
Comment     float64
dtype: object


In [9]:
print("--- sanity: Sanity value counts ---")
print(dfs["sanity"]["Sanity"].value_counts())
print()

print("--- fh_rsl: Status value counts ---")
print(dfs["fh_rsl"]["Status"].value_counts())

--- sanity: Sanity value counts ---
Sanity
PREVENTIVE    2571
CURATIVE      2352
OK            1115
Name: count, dtype: int64

--- fh_rsl: Status value counts ---
Status
Lien_OK                855
Lien_dépointé_(<10)    238
Lien_dépointé_(>10)    199
Name: count, dtype: int64


In [10]:
print("--- fh_rsl numeric summary ---")
display(dfs["fh_rsl"][["Max RSL", "RSL REF", "RSL DIFF"]].describe())

print("--- rtwp numeric summary ---")
display(dfs["rtwp"][["mean_rtwp", "max_rtwp", "min_rtwp", "occ_rtwp_gt_98",
                      "vs_mean_rtwp_ant0", "vs_mean_rtwp_ant1", "occ_diff_sect_gt_5"]].describe())

--- fh_rsl numeric summary ---


,Max RSL,RSL REF,RSL DIFF
count,1292.000000,1292.000000,1292.000000
mean,-31.364551,-27.882817,3.481734
std,10.689109,7.666919,11.720095
min,-99.000000,-46.860000,-38.740000
25%,-35.125000,-32.350000,-2.110000
50%,-30.700000,-29.740000,2.190000
75%,-26.800000,-25.177500,7.170000
max,0.000000,8.600000,94.310000


--- rtwp numeric summary ---


,mean_rtwp,max_rtwp,min_rtwp,occ_rtwp_gt_98,vs_mean_rtwp_ant0,vs_mean_rtwp_ant1,occ_diff_sect_gt_5
count,295.000000,295.000000,295.000000,295.000000,295.000000,295.000000,295.000000
mean,-98.072053,-89.204900,-102.572620,31.142373,-98.040326,-97.794164,13.203390
std,3.125526,5.556855,3.796924,17.755522,3.751665,3.478424,19.925354
min,-103.844000,-96.905100,-107.633000,11.000000,-106.529000,-105.110000,0.000000
25%,-100.111500,-93.157600,-105.306500,16.000000,-100.268500,-99.738950,0.000000
50%,-98.684200,-90.216300,-103.781000,26.000000,-98.633800,-98.418300,0.000000
75%,-97.030600,-86.686900,-100.225000,44.500000,-96.236350,-96.194100,23.500000
max,-76.227800,-69.216500,-80.019800,63.000000,-74.855500,-77.965800,63.000000


In [11]:
print("--- rows with Max RSL == 0 ---")
display(dfs["fh_rsl"][dfs["fh_rsl"]["Max RSL"] == 0])

print("--- rows with Max RSL == -99 (or very close) ---")
display(dfs["fh_rsl"][dfs["fh_rsl"]["Max RSL"] <= -98.5])

print("--- top 5 highest RSL DIFF ---")
display(dfs["fh_rsl"].nlargest(5, "RSL DIFF"))

print("--- top 5 lowest RSL DIFF ---")
display(dfs["fh_rsl"].nsmallest(5, "RSL DIFF"))

--- rows with Max RSL == 0 ---


,Status,IP,Max RSL,RSL REF,RSL DIFF,Name,File,Comment
7,Lien_OK,172.18.250.6,0.0,-26.97,-26.97,B2B_SOTRAMI_KEF_0022,SOTRAMI_KEF_0022.xlsx,NaN
32,Lien_OK,172.18.152.26,0.0,-16.73,-16.73,B2B_IMSET_GABES_GAB_0067,IMSET_GABES - GAB_0067_C01.xlsx,NaN
41,Lien_OK,172.19.184.12,0.0,-22.93,-22.93,B2B_SNBG-GABES_GAB_0016,SNBG_GAB_0016.xlsx,NaN
50,Lien_OK,172.19.104.17,0.0,-23.40,-23.40,B2B_TRS_TUNISIE_TUN_0106,TRS_TUNISIE - TUN_0106_C03.xlsx,NaN
53,Lien_OK,172.19.206.7,0.0,-26.55,-26.55,B2B_ATB SBO_SBO_0010,ATB_BOUZID_SBO_0010.xlsx,NaN
71,Lien_OK,172.18.38.4,0.0,-35.00,-35.00,B2B_CHARK_TABAC_MON_0011,CHARK_TABAC - MON_0011_C05.xlsx,NaN
86,Lien_OK,172.18.40.99,0.0,-27.45,-27.45,B2B_MONOPRIX MERCURE_SOU_0004,Monoprix_mercure_sou_SOU_0004.xlsx,NaN
94,Lien_OK,172.18.134.47,0.0,-30.33,-30.33,B2B_MODERN_METAL_INTER_MON_0045,B2B_MODERN_METAL_MON_0017.xlsx,NaN
112,Lien_OK,172.18.38.12,0.0,-30.81,-30.81,B2B_BEYTIBRICO_MON_0011,B2B BEYTIBRICO v MON_0011.xlsx,NaN
118,Lien_OK,172.19.208.11,0.0,-32.09,-32.09,B2B_FRANCHISE_TAT._TAT_0009,FRANCHISE_TATAOUINE - TAT_0009_C01.xlsx,NaN


--- rows with Max RSL == -99 (or very close) ---


,Status,IP,Max RSL,RSL REF,RSL DIFF,Name,File,Comment
853,Lien_dépointé_(>10),172.19.188.3,-99.0,-34.07,64.93,TUN_0001_FOUNA_FH_0360_NINGEN,NINGEN_TUN_0001.xlsx,NaN


--- top 5 highest RSL DIFF ---


,Status,IP,Max RSL,RSL REF,RSL DIFF,Name,File,Comment
858,Lien_dépointé_(>10),172.19.236.27,-97.4,-3.09,94.31,KEF_0010_BT DAHMANI,BT_DAHMENI_KEF_0010.xlsx,NaN
863,Lien_dépointé_(>10),172.18.40.102,-95.4,-18.44,76.96,SOU_0004_HERSAND,HERSAND - SOU_0004_C01.xlsx,NaN
319,Lien_dépointé_(>10),172.18.226.1,-95.7,-20.01,75.69,TUN_0284_GAT_ASSURANCE_EXPERTISE,B2B GAT_ASSURANCE_EXPERTISE - TUN_0284_C02.xlsx,NaN
867,Lien_dépointé_(>10),172.18.140.7,-92.9,-24.89,68.01,TUN_0011_ZITOUNA,ZITOUNA_LAC_LAC_TUN_0011.xlsx,NaN
854,Lien_dépointé_(>10),172.19.199.65,-97.9,-30.73,67.17,NAB_0033_UIB_NABEUL,MLO_UIB_NABEUL_RT_HAMMAMET v NAB_0033.xlsx,NaN


--- top 5 lowest RSL DIFF ---


,Status,IP,Max RSL,RSL REF,RSL DIFF,Name,File,Comment
729,Lien_OK,172.18.3.130,0.0,-38.74,-38.74,SOU052-SOU007&STS,STA_SOU_0078.xlsx,NaN
71,Lien_OK,172.18.38.4,0.0,-35.00,-35.00,B2B_CHARK_TABAC_MON_0011,CHARK_TABAC - MON_0011_C05.xlsx,NaN
625,Lien_OK,172.18.38.9,0.0,-33.35,-33.35,MON_0011_KARA INDUSTRIE,KARA_INDUSTRIE - MON_0011_C05.xlsx,NaN
139,Lien_OK,172.19.38.9,0.0,-32.38,-32.38,B2B_CALL CENTRE SIEGE_SOU_0091,PB_CALL_CENTER_Siege_SOU_0091.xlsx,NaN
118,Lien_OK,172.19.208.11,0.0,-32.09,-32.09,B2B_FRANCHISE_TAT._TAT_0009,FRANCHISE_TATAOUINE - TAT_0009_C01.xlsx,NaN


In [12]:
# rxlev: check for the same -99.90 sentinel pattern (as strings, since not yet converted)
print("--- rxlev: Min RSL value counts (top 10) ---")
print(dfs["rxlev"]["Min RSL"].value_counts().head(10))

# rtwp: check occ_rtwp_gt_98 and occ_diff_sect_gt_5 for zero-inflation (many rows at 0?)
print("\n--- rtwp: occ_rtwp_gt_98 distribution ---")
print(dfs["rtwp"]["occ_rtwp_gt_98"].describe())
print("\n--- rtwp: occ_diff_sect_gt_5 distribution ---")
print(dfs["rtwp"]["occ_diff_sect_gt_5"].describe())

--- rxlev: Min RSL value counts (top 10) ---
Min RSL
-99.90 dBm    709
-99.00 dBm    151
-96.00 dBm     33
-96.40 dBm     32
-96.10 dBm     32
-95.80 dBm     31
-94.30 dBm     31
-94.60 dBm     30
-95.70 dBm     29
-95.00 dBm     29
Name: count, dtype: int64

--- rtwp: occ_rtwp_gt_98 distribution ---
count    295.000000
mean      31.142373
std       17.755522
min       11.000000
25%       16.000000
50%       26.000000
75%       44.500000
max       63.000000
Name: occ_rtwp_gt_98, dtype: float64

--- rtwp: occ_diff_sect_gt_5 distribution ---
count    295.000000
mean      13.203390
std       19.925354
min        0.000000
25%        0.000000
50%        0.000000
75%       23.500000
max       63.000000
Name: occ_diff_sect_gt_5, dtype: float64


the MLO work 

In [13]:
mlo_dir = DATA_DIR / "MLO"
mlo_files = list(mlo_dir.glob("*.xlsx"))
print("total MLO files found:", len(mlo_files))
print("first 5 filenames:")
for f in mlo_files[:5]:
    print(" -", f.name)

total MLO files found: 3961
first 5 filenames:
 - 2AP_CONSEIL_Tunis - ARI_0012_C01.xlsx
 - ABARAKA_Palmarium - SFA_0040_C01.xlsx
 - ABATTOIR_CHAHIA_Zarzis - MED_0025_C03.xlsx
 - ABT_TUN_0009.xlsx
 - ADP_CHARGUIA_2 - TUN_0091_C02.xlsx


In [14]:
import openpyxl

sample_files = mlo_files[:5]

for f in sample_files:
    wb = openpyxl.load_workbook(f, read_only=True, data_only=True)
    print(f"--- {f.name} ---")
    print("sheet names:", wb.sheetnames)
    for sheet in wb.sheetnames:
        ws = wb[sheet]
        print(f"  sheet '{sheet}': max_row = {ws.max_row}, max_col = {ws.max_column}")
    wb.close()
    print()

--- 2AP_CONSEIL_Tunis - ARI_0012_C01.xlsx ---
sheet names: ['Feuil1', 'Feuil2', 'Feuil3']
  sheet 'Feuil1': max_row = 71, max_col = 5
  sheet 'Feuil2': max_row = 1, max_col = 1
  sheet 'Feuil3': max_row = 1, max_col = 1

--- ABARAKA_Palmarium - SFA_0040_C01.xlsx ---
sheet names: ['Feuil1', 'Feuil2', 'Feuil3']
  sheet 'Feuil1': max_row = 71, max_col = 5
  sheet 'Feuil2': max_row = 1, max_col = 1
  sheet 'Feuil3': max_row = 1, max_col = 1

--- ABATTOIR_CHAHIA_Zarzis - MED_0025_C03.xlsx ---
sheet names: ['Feuil1', 'Feuil2', 'Feuil3']
  sheet 'Feuil1': max_row = 71, max_col = 5
  sheet 'Feuil2': max_row = 1, max_col = 1
  sheet 'Feuil3': max_row = 1, max_col = 1

--- ABT_TUN_0009.xlsx ---
sheet names: ['Feuil1', 'Feuil2', 'Feuil3']
  sheet 'Feuil1': max_row = 71, max_col = 5
  sheet 'Feuil2': max_row = 1, max_col = 1
  sheet 'Feuil3': max_row = 1, max_col = 1

--- ADP_CHARGUIA_2 - TUN_0091_C02.xlsx ---
sheet names: ['Feuil1', 'Feuil2', 'Feuil3']
  sheet 'Feuil1': max_row = 71, max_col = 5


In [15]:
sample = mlo_files[0]
df_sample = pd.read_excel(sample, sheet_name="Feuil1", header=None)
print(f"--- {sample.name} : Feuil1 raw content ---")
print("shape:", df_sample.shape)
pd.set_option("display.max_rows", 80)
display(df_sample)

--- 2AP_CONSEIL_Tunis - ARI_0012_C01.xlsx : Feuil1 raw content ---
shape: (70, 4)


,0,1,2,3
0,NaN,Atoll: Link Reporter,NaN,NaN
1,NaN,Generation date : lundi 30 mai 2022 20:16:48,NaN,NaN
2,NaN,NaN,NaN,NaN
3,NaN,MLO 2AP_CONSEIL_Tunis - ARI_0012_C01,NaN,NaN
4,NaN,NaN,NaN,NaN
5,NaN,NaN,NaN,NaN
6,LINK SPECIFICATION,NaN,NaN,NaN
7,NaN,NaN,NaN,NaN
8,NaN,NaN,NaN,NaN
9,NaN,NaN,NaN,NaN


In [16]:
import random
random.seed(42)
random_sample = random.sample(mlo_files, 15)

for f in random_sample:
    df = pd.read_excel(f, sheet_name="Feuil1", header=None)
    # find the row where column 0 == "Link ID" isn't right since Link ID is in col... let's search for known labels
    link_id_row = df[df[1] == "Link ID"].index.tolist()
    freq_band_row = df[df[1] == "Band"].index.tolist()
    print(f"{f.name}: shape={df.shape}, 'Link ID' row={link_id_row}, 'Band' row={freq_band_row}")

OLA_ENERGY_RADES - BAR_0127_C02.xlsx: shape=(70, 4), 'Link ID' row=[], 'Band' row=[35]
BANSARD_TUNIS - BAR_0078_C01.xlsx: shape=(70, 4), 'Link ID' row=[], 'Band' row=[35]
ARI_0070_ARI_0092_LTE_V.xlsx: shape=(70, 4), 'Link ID' row=[], 'Band' row=[35]
SFA_0201_SFA_0001_LTE_T_V.xlsx: shape=(70, 4), 'Link ID' row=[], 'Band' row=[35]
GAB_0045_GAB_0027_H.xlsx: shape=(70, 4), 'Link ID' row=[], 'Band' row=[35]
EL_KSAR_01 - SBO_0028_C01.xlsx: shape=(70, 4), 'Link ID' row=[], 'Band' row=[35]
CIMENT_ARTIFICIEL_TUN_0105.xlsx: shape=(70, 4), 'Link ID' row=[], 'Band' row=[35]
BAR_0153_BAR_0072_V.xlsx: shape=(70, 4), 'Link ID' row=[], 'Band' row=[35]
SFA_0179_SFA_0101_H_Upg_Final.xlsx: shape=(70, 4), 'Link ID' row=[], 'Band' row=[35]
B2B STE_LES_MOULINS_ERRAYEN v BIZ_0042.xlsx: shape=(70, 4), 'Link ID' row=[], 'Band' row=[35]
SBO_0035_SBO_0041_LTE_V.xlsx: shape=(70, 4), 'Link ID' row=[], 'Band' row=[35]
SFA_0198_SFA_0102_LTE_V.xlsx: shape=(70, 4), 'Link ID' row=[], 'Band' row=[35]
TUN_0283_TUN_0111.x

In [17]:
for f in random_sample:
    df = pd.read_excel(f, sheet_name="Feuil1", header=None)
    link_id_row = df[df[0] == "Link ID"].index.tolist()
    capacity_row = df[df[0] == "Capacity"].index.tolist()
    band_row = df[df[1] == "Band"].index.tolist()
    tx_power_row = df[df[1] == "Tx Power (dBm)"].index.tolist()
    threshold_row = df[df[1] == "Threshold"].index.tolist()
    print(f"{f.name}: Link ID={link_id_row}, Capacity={capacity_row}, Band={band_row}, TxPower={tx_power_row}, Threshold={threshold_row}")

OLA_ENERGY_RADES - BAR_0127_C02.xlsx: Link ID=[25], Capacity=[26], Band=[35], TxPower=[53], Threshold=[62]
BANSARD_TUNIS - BAR_0078_C01.xlsx: Link ID=[25], Capacity=[26], Band=[35], TxPower=[53], Threshold=[62]
ARI_0070_ARI_0092_LTE_V.xlsx: Link ID=[25], Capacity=[26], Band=[35], TxPower=[53], Threshold=[62]
SFA_0201_SFA_0001_LTE_T_V.xlsx: Link ID=[25], Capacity=[26], Band=[35], TxPower=[53], Threshold=[62]
GAB_0045_GAB_0027_H.xlsx: Link ID=[25], Capacity=[26], Band=[35], TxPower=[53], Threshold=[62]
EL_KSAR_01 - SBO_0028_C01.xlsx: Link ID=[25], Capacity=[26], Band=[35], TxPower=[53], Threshold=[62]
CIMENT_ARTIFICIEL_TUN_0105.xlsx: Link ID=[25], Capacity=[26], Band=[35], TxPower=[53], Threshold=[62]
BAR_0153_BAR_0072_V.xlsx: Link ID=[25], Capacity=[26], Band=[35], TxPower=[53], Threshold=[62]
SFA_0179_SFA_0101_H_Upg_Final.xlsx: Link ID=[25], Capacity=[26], Band=[35], TxPower=[53], Threshold=[62]
B2B STE_LES_MOULINS_ERRAYEN v BIZ_0042.xlsx: Link ID=[25], Capacity=[26], Band=[35], TxPowe

In [18]:
import sys
sys.path.append(str(Path("../src").resolve()))
from preprocessing.mlo_parser import parse_mlo_file

results = [parse_mlo_file(f) for f in random_sample]
df_test = pd.DataFrame(results)
display(df_test)

,_file,_error,link_id,capacity,length_m,calc_method,band_end_a,band_end_b,bandwidth_mhz_end_a,bandwidth_mhz_end_b,...,modulation_end_a,modulation_end_b,threshold_dbm_end_a,threshold_dbm_end_b,main_rx_level_dbm_end_a,main_rx_level_dbm_end_b,threshold_degradation_db_end_a,threshold_degradation_db_end_b,main_flat_fade_margin_db_end_a,main_flat_fade_margin_db_end_b
0,OLA_ENERGY_RADES - BAR_0127_C02.xlsx,None,OLA_ENERGY_RADES - BAR_0127_C02,"10,2 Mbps","445,24 m",ITU-R P. 530-12,38 GHz,38 GHz,LK_38GHz_3.5MHz[-],LK_38GHz_3.5MHz[+],...,16QAM,NaN,"-83,5 dBm","-83,5 dBm","-30,97 dBm","-31,27 dBm",3 dB,3 dB,"52,23 dB","52,53 dB"
1,BANSARD_TUNIS - BAR_0078_C01.xlsx,None,BANSARD_TUNIS - BAR_0078_C01,"10,2 Mbps","2 023,68 m",ITU-R P. 530-12,38 GHz,38 GHz,LK_38GHz_3.5MHz[-],LK_38GHz_3.5MHz[+],...,16QAM,NaN,"-83,5 dBm","-83,5 dBm","-30,82 dBm","-31,13 dBm",3 dB,3 dB,"52,37 dB","52,68 dB"
2,ARI_0070_ARI_0092_LTE_V.xlsx,None,ARI_0070_ARI_0092_LTE_V,222 Mbps,"4 122,25 m",ITU-R P. 530-12,23 GHz,23 GHz,HD_23GHz_28MHz[-],HD_23GHz_28MHz[+],...,1024QAM,NaN,-61 dBm,-61 dBm,"-38,84 dBm","-39,22 dBm",3 dB,3 dB,"21,78 dB","22,16 dB"
3,SFA_0201_SFA_0001_LTE_T_V.xlsx,None,SFA_0201_SFA_0001_LTE_T_V,318 Mbps,"2 505,54 m",ITU-R P. 530-12,23 GHz,23 GHz,HE2_23GHz_56MHz[+],HE2_23GHz_56MHz[-],...,128QAM,NaN,-68 dBm,-68 dBm,"-31,44 dBm","-31,06 dBm",3 dB,3 dB,"36,94 dB","36,56 dB"
4,GAB_0045_GAB_0027_H.xlsx,None,GAB_0045_GAB_0027_H,407 Mbps,"13 143,27 m",ITU-R P. 530-12,11 GHz,11 GHz,BW_11GHz_56MHz[+],BW_11GHz_56MHz[-],...,512QAM,NaN,"-62,5 dBm","-62,5 dBm","-34,74 dBm","-34,31 dBm",3 dB,3 dB,"28,19 dB","27,76 dB"
5,EL_KSAR_01 - SBO_0028_C01.xlsx,None,EL_KSAR_01 - SBO_0028_C01,"20,5 Mbps","10 957,91 m",ITU-R P. 530-12,23 GHz,23 GHz,HF_23GHz_7MHz[-],HF_23GHz_7MHz[+],...,16QAM,NaN,-84 dBm,-84 dBm,"-39,3 dBm","-39,3 dBm",3 dB,3 dB,"44,7 dB","44,7 dB"
6,CIMENT_ARTIFICIEL_TUN_0105.xlsx,None,CIMENT_ARTIFICIEL_TUN_0105,"20,5 Mbps","2 152,93 m",ITU-R P. 530-12,38 GHz,38 GHz,LH_38GHz_14MHz[-],LH_38GHz_14MHz[+],...,Q-PSK,NaN,"-87,5 dBm","-87,5 dBm","-23,88 dBm","-24,18 dBm",3 dB,3 dB,"63,32 dB","63,62 dB"
7,BAR_0153_BAR_0072_V.xlsx,None,BAR_0153_BAR_0072_V,247 Mbps,"1 824,15 m",ITU-R P. 530-12,38 GHz,38 GHz,LL_38GHz_28MHz[-],LL_38GHz_28MHz[+],...,2048QAM,NaN,-55 dBm,-55 dBm,"-31,63 dBm","-32,02 dBm",3 dB,3 dB,"22,98 dB","23,37 dB"
8,SFA_0179_SFA_0101_H_Upg_Final.xlsx,None,SFA_0179_SFA_0101_H,495 Mbps,"8 921,03 m",ITU-R P. 530-12,18 GHz,18 GHz,GC_18GHz_55MHz[-],GC_18GHz_55MHz[+],...,2048QAM,NaN,-54 dBm,-54 dBm,"-23,67 dBm","-24,43 dBm",1 dB,1 dB,"29,57 dB","30,33 dB"
9,B2B STE_LES_MOULINS_ERRAYEN v BIZ_0042.xlsx,None,STE_LES_MOULINS_ERRAYEN_BIZ_0042,"31,8 Mbps","1 309,81 m",ITU-R P. 530-12,38 GHz,38 GHz,LI_38GHz_7MHz[+],LI_38GHz_7MHz[-],...,64QAM,NaN,-78 dBm,-78 dBm,"-27,7 dBm","-27,39 dBm",3 dB,3 dB,"50,61 dB","50,3 dB"


In [19]:
import importlib
import preprocessing.mlo_parser
importlib.reload(preprocessing.mlo_parser)
from preprocessing.mlo_parser import parse_mlo_file

In [20]:
all_results = [parse_mlo_file(f) for f in mlo_files]
df_mlo = pd.DataFrame(all_results)

print("total parsed:", len(df_mlo))
print("errors:", df_mlo["_error"].notna().sum())
print()
print("rows with errors:")
display(df_mlo[df_mlo["_error"].notna()])

total parsed: 3961
errors: 166

rows with errors:


,_file,_error,link_id,capacity,length_m,calc_method,band_end_a,band_end_b,bandwidth_mhz_end_a,bandwidth_mhz_end_b,...,modulation_end_a,modulation_end_b,threshold_dbm_end_a,threshold_dbm_end_b,main_rx_level_dbm_end_a,main_rx_level_dbm_end_b,threshold_degradation_db_end_a,threshold_degradation_db_end_b,main_flat_fade_margin_db_end_a,main_flat_fade_margin_db_end_b
1038,E_Band.xlsx,"unexpected shape (0, 0)",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3714,Upgrade JEN_0030_JEN_0005.xlsx,"unexpected shape (28, 1)",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3797,~$2AP_CONSEIL_Tunis - ARI_0012_C01.xlsx,read failed: Excel file format cannot be deter...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3798,~$ALTEA-PACK_Main_SFA_0001.xlsx,read failed: Excel file format cannot be deter...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3799,~$ARI_0032_TUN_0099_H.xlsx,read failed: Excel file format cannot be deter...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3956,~$ZAG_0014_ZAG_0008_LTE_H.xlsx,read failed: Excel file format cannot be deter...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3957,~$ZAG_0014_ZAG_0008_LTE_V.xlsx,read failed: Excel file format cannot be deter...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3958,~$ZAG_0117_KAI_0030_1_H.xlsx,read failed: Excel file format cannot be deter...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3959,~$ZAG_0133_ZAG_0017.xlsx,read failed: Excel file format cannot be deter...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [21]:
print(open(Path("../src/preprocessing/mlo_parser.py")).read())

import pandas as pd
from pathlib import Path

# (row, column) positions confirmed fixed across 20 sampled files
FIELD_MAP = {
    "link_id":        (25, 0, "value_direct"),   # label in col0, value in col1
    "capacity":       (26, 0, "value_direct"),
    "length_m":       (25, 2, "value_direct"),   # label in col2, value in col3
    "calc_method":    (26, 2, "value_direct"),
    "band":           (35, 1, "end_ab"),         # label col1, values col2/col3
    "bandwidth_mhz":  (38, 1, "end_ab"),
    "polarisation":   (40, 1, "end_ab"),
    "antenna_type":   (42, 1, "end_ab"),
    "antenna_size_m": (43, 1, "end_ab"),
    "antenna_height_m": (44, 1, "end_ab"),
    "antenna_direction_deg": (45, 1, "end_ab"),
    "antenna_gain_db": (47, 1, "end_ab"),
    "tx_power_dbm":   (53, 1, "end_ab"),
    "tx_attenuator_db": (54, 1, "end_ab"),
    "rx_attenuator_db": (55, 1, "end_ab"),
    "modulation":     (56, 1, "end_ab"),
    "threshold_dbm":  (62, 1, "end_ab"),
    "main_rx_level_dbm": (65, 1, "

In [22]:
all_results = [parse_mlo_file(f) for f in mlo_files]
df_mlo = pd.DataFrame(all_results)

print("total parsed:", len(df_mlo))
print("errors:", df_mlo["_error"].notna().sum())
print()
print("error breakdown:")
print(df_mlo[df_mlo["_error"].notna()]["_error"].value_counts())

total parsed: 3961
errors: 166

error breakdown:
_error
read failed: Excel file format cannot be determined, you must specify an engine manually.    164
unexpected shape (0, 0)                                                                        1
unexpected shape (28, 1)                                                                       1
Name: count, dtype: int64


In [23]:
# filter out Excel lock files before parsing
mlo_files_clean = [f for f in mlo_files if not f.name.startswith("~$")]
print("files after filtering lock files:", len(mlo_files_clean))
print("lock files excluded:", len(mlo_files) - len(mlo_files_clean))

files after filtering lock files: 3797
lock files excluded: 164


In [24]:
all_results = [parse_mlo_file(f) for f in mlo_files_clean]
df_mlo = pd.DataFrame(all_results)

print("total parsed:", len(df_mlo))
print("errors:", df_mlo["_error"].notna().sum())
print()
print(df_mlo[df_mlo["_error"].notna()])

total parsed: 3797
errors: 2

                               _file                    _error link_id  \
1038                     E_Band.xlsx   unexpected shape (0, 0)     NaN   
3714  Upgrade JEN_0030_JEN_0005.xlsx  unexpected shape (28, 1)     NaN   

     capacity length_m calc_method band_end_a band_end_b bandwidth_mhz_end_a  \
1038      NaN      NaN         NaN        NaN        NaN                 NaN   
3714      NaN      NaN         NaN        NaN        NaN                 NaN   

     bandwidth_mhz_end_b  ... modulation_end_a modulation_end_b  \
1038                 NaN  ...              NaN              NaN   
3714                 NaN  ...              NaN              NaN   

     threshold_dbm_end_a threshold_dbm_end_b main_rx_level_dbm_end_a  \
1038                 NaN                 NaN                     NaN   
3714                 NaN                 NaN                     NaN   

     main_rx_level_dbm_end_b threshold_degradation_db_end_a  \
1038                    

In [ ]:
df_mlo_clean = df_mlo[df_mlo["_error"].isna()].drop(columns=["_error"]).reset_index(drop=True)

print("clean rows:", df_mlo_clean.shape)
display(df_mlo_clean.head())

output_path = Path("../data/processed/atoll_links.csv")
output_path.parent.mkdir(parents=True, exist_ok=True)
df_mlo_clean.to_csv(output_path, index=False)
print("saved to:", output_path.resolve())

df_mlo_errors = df_mlo[df_mlo["_error"].notna()]
df_mlo_errors.to_csv(Path("../data/processed/atoll_links_parse_errors.csv"), index=False)
print("excluded files logged to atoll_links_parse_errors.csv")

clean rows: (3795, 37)


,_file,link_id,capacity,length_m,calc_method,band_end_a,band_end_b,bandwidth_mhz_end_a,bandwidth_mhz_end_b,polarisation_end_a,...,modulation_end_a,modulation_end_b,threshold_dbm_end_a,threshold_dbm_end_b,main_rx_level_dbm_end_a,main_rx_level_dbm_end_b,threshold_degradation_db_end_a,threshold_degradation_db_end_b,main_flat_fade_margin_db_end_a,main_flat_fade_margin_db_end_b
0,2AP_CONSEIL_Tunis - ARI_0012_C01.xlsx,2AP_CONSEIL_Tunis - ARI_0012_C01,41 Mbps,"700,77 m",ITU-R P. 530-12,23 GHz,23 GHz,HC_23GHz_14MHz[+],HC_23GHz_14MHz[-],Horizontal,...,16QAM,NaN,"-78,5 dBm","-78,5 dBm","-25,07 dBm","-24,69 dBm",3 dB,3 dB,"53,81 dB","53,43 dB"
1,ABARAKA_Palmarium - SFA_0040_C01.xlsx,ABARAKA_Palmarium - SFA_0040_C01,"10,2 Mbps","986,01 m",ITU-R P. 530-12,38 GHz,38 GHz,LK_38GHz_3.5MHz[+],LK_38GHz_3.5MHz[-],Vertical,...,16QAM,NaN,"-83,5 dBm","-83,5 dBm","-30,27 dBm","-29,97 dBm",3 dB,3 dB,"53,53 dB","53,23 dB"
2,ABATTOIR_CHAHIA_Zarzis - MED_0025_C03.xlsx,ABATTOIR_CHAHIA_Zarzis - MED_0025_C03,"20,5 Mbps","5 205,26 m",ITU-R P. 530-12,23 GHz,23 GHz,HF_23GHz_7MHz[+],HF_23GHz_7MHz[-],Vertical,...,16QAM,NaN,-84 dBm,-84 dBm,"-38,26 dBm","-37,89 dBm",3 dB,3 dB,"46,11 dB","45,74 dB"
3,ABT_TUN_0009.xlsx,ABT_TUN_0009,203 Mbps,"314,69 m",ITU-R P. 530-12,38 GHz,38 GHz,LL_38GHz_28MHz[+],LL_38GHz_28MHz[-],Vertical,...,512QAM,NaN,-63 dBm,-63 dBm,"-29,72 dBm","-29,43 dBm",1 dB,1 dB,"33,57 dB","33,28 dB"
4,ADP_CHARGUIA_2 - TUN_0091_C02.xlsx,ADP_CHARGUIA_2 - TUN_0091_C02,14 Mbps,"1 506,39 m",ITU-R P. 530-12,38 GHz,38 GHz,LI_38GHz_7MHz[-],LI_38GHz_7MHz[+],Horizontal,...,16QAM,NaN,-84 dBm,-84 dBm,"-22,17 dBm","-22,52 dBm",3 dB,3 dB,"61,48 dB","61,83 dB"


saved to: C:\OrangeCopilot\projet_pfe-main (1)\projet_pfe-main\main\FH\AI\data\processed\atoll_links.csv
excluded files logged to atoll_links_parse_errors.csv


# 01 — Data Overview & MLO Parsing

**Goal:** Initial exploration of the 4 core telecom datasets (SANITY, RXLev, FH RSL, RTWP),
plus building and running the parser for the ~3961 Atoll Link Reporter (MLO) files.

**Inputs:** `FH/data/*.xlsx`, `FH/data/temp/*.xlsx`, `FH/data/MLO/*.xlsx`

**Outputs:**
- Confirmed structure, dtypes, missing values, duplicates for all 4 core datasets
- Identified `Sanity` and `Status` as classification targets
- Identified sentinel-value pattern (0 dBm / ~-99 dBm = "no reading")
- `src/preprocessing/mlo_parser.py` — parses one Atoll Link Reporter file into a flat dict
- `data/processed/atoll_links.csv` — 3795 parsed MLO links
- `data/processed/atoll_links_parse_errors.csv` — 166 excluded files (164 Excel lock files, 2 malformed)